src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -28,21 +27,21 @@ namespace Microsoft.AspNet.Mvc.Razor
     public abstract class RazorPage : IRazorPage
     {
         private readonly HashSet<string> _renderedSections = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-        private readonly Stack<TextWriter> _writerScopes;
-        private TextWriter _originalWriter;
+        private readonly Stack<IHtmlContentBuilder> _contentBuilderScopes;
+        private IHtmlContentBuilder _originalWriter;
         private IUrlHelper _urlHelper;
         private ITagHelperActivator _tagHelperActivator;
         private ITypeActivatorCache _typeActivatorCache;
         private bool _renderedBody;
         private AttributeInfo _attributeInfo;
         private TagHelperAttributeInfo _tagHelperAttributeInfo;
-        private StringCollectionTextWriter _valueBuffer;
+        private HtmlContentBuilder _valueBuffer;
```

Comment:
This is going to get re-specialized when we integrate pooling here. No reason to block you, but FYI.